src/lib/asset-proxy.js

```diff
@@ -0,0 +1,338 @@
+        const pathParts = pathname.split('/').filter((p) => p)
+
+        if (pathParts.length < 2) {
+          return null // Need at least bucket and key
+        }
+
+        const bucket = pathParts[0]
+        const key = pathParts.slice(1).join('/')
+
+        return { bucket, key, region }
+      }
+    } catch (_error) {
+      // Invalid URL
+      return null
+    }
+  }
+
+  return null
+}
+
+/**
+ * @param {Object} asset - Asset object
+ * @param {Object} itemOrCollection - Item or Collection object
+ * @returns {string} AWS region
+ */
+const determineS3Region = (asset, itemOrCollection) => {
+  // Storage Extension v1
+  const v1Region = asset['storage:region'] || itemOrCollection.properties?.['storage:region']
+  if (v1Region) {
+    return v1Region
+  }
+
+  // Storage Extension v2
+  const storageSchemes = itemOrCollection.properties?.['storage:schemes']
+    || itemOrCollection['storage:schemes']
+  const v2Region = storageSchemes?.[asset['storage:refs']]?.region
+  if (v2Region) {
+    return v2Region
+  }
+
+  return process.env['AWS_REGION'] || 'us-west-2'
+}
+
+export class AssetProxy {
+  constructor() {
+    this.bucketsCache = null
+    this.s3ClientCache = new Map()
+    this.bucketOption = process.env['ASSET_PROXY_BUCKET_OPTION'] || 'NONE'
+    this.bucketList = process.env['ASSET_PROXY_BUCKET_LIST']
+    this.urlExpiry = parseInt(process.env['ASSET_PROXY_URL_EXPIRY'] || '300', 10)
+  }
+
+  /**
+   * @returns {Promise<void>}
+   */
+  async initialize() {
+    switch (this.bucketOption) {
+    case BucketOption.LIST:
+      if (this.bucketList) {
+        const bucketNames = this.bucketList.split(',').map((b) => b.trim()).filter((b) => b)
+        this.bucketsCache = new Set(bucketNames)
+        logger.info(
+          `Parsed ${this.bucketsCache.size} buckets from ASSET_PROXY_BUCKET_LIST for asset proxy`
+        )
+      } else {
+        throw new Error(
+          'ASSET_PROXY_BUCKET_LIST must be set when ASSET_PROXY_BUCKET_OPTION is LIST'
+        )
+      }
+      break
+
+    case BucketOption.ALL_BUCKETS_IN_ACCOUNT:
+      try {
+        const region = process.env['AWS_REGION'] || 'us-west-2'
+        const client = this.getS3Client(region)
+        const command = new ListBucketsCommand({})
+        const response = await client.send(command)
+        const bucketNames = response.Buckets?.map((b) => b.Name)
+          ?.filter((name) => typeof name === 'string') || []
+        this.bucketsCache = new Set(bucketNames)
+        logger.info(`Fetched ${this.bucketsCache.size} buckets from AWS account for asset proxy`)
+      } catch (error) {
+        const message = error instanceof Error ? error.message : String(error)
+        throw new Error(`Failed to fetch buckets for asset proxy: ${message}`)
+      }
+      break
+
+    default:
+      break
+    }
+  }
+
+  /**
+   * @param {string} region - AWS region
+   * @returns {Object} S3 client instance
+   */
+  getS3Client(region) {
```

Comment:
I'm not sure you need to do this whole having a different client per region. Looking at https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/Package/-aws-sdk-s3-request-presigner/Function/getSignedUrl3/ it appears you can pass a `signingRegion` parameter, which I believe would mean you need only one S3 client for all operations.

If you need only one S3 client then you want to create it during initialization, so that can be run as part of the app initialization within the lambda handler init.

Reply:
Excellent suggestion. Done.